xapian-applications/omega/handler_gmime.cc

```diff
@@ -0,0 +1,173 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "myhtmlparse.h"
+#include "utf8convert.h"
+#include <string.h>
+
+#include <glib.h>
+#include <gmime/gmime.h>
+
+using namespace std;
+
+const unsigned SIZE = 4096;
+
+static void
+extract_html(const string& text, string& charset, string& dump)
+{
+    MyHtmlParser parser;
+    if (charset.empty())
+	charset = "UTF-8";
+    try {
+	parser.ignore_metarobots();
+	parser.parse_html(text, charset, false);
+	dump += parser.dump;
+    } catch (const string& newcharset) {
+	parser.reset();
+	parser.ignore_metarobots();
+	parser.parse_html(text, newcharset, true);
+	dump += parser.dump;
+    } catch (...) {
+	return;
+    }
+}
+
+size_t decode(unsigned char* text, size_t len, GMimeContentEncoding encoding)
+{
+    unsigned char buffer[SIZE];
+    static int state = 0;
+    static guint32 save = 0;
+    switch (encoding) {
+	case GMIME_CONTENT_ENCODING_BASE64: {
+	    len = g_mime_encoding_base64_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_UUENCODE: {
+	    len = g_mime_encoding_uudecode_step(text, len, buffer,
+						&state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_QUOTEDPRINTABLE: {
+	    len = g_mime_encoding_quoted_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	default:
+	    break;
+    }
+    return len;
+}
+
+bool
+parser_content(GMimeObject* me, string& dump)
+{
+    GMimeContentType* ct = g_mime_object_get_content_type(me);
+    if (GMIME_IS_MULTIPART(me)) {
+	GMimeMultipart* mpart = (GMimeMultipart*)me;
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	int count = g_mime_multipart_get_count(mpart);
+	if (subtype == "alternative") {
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		if (parser_content(part, dump))
+		    return true;
+	    }
+	} else {
+	    bool ret = false;
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		ret |= parser_content(part, dump);
+	    }
+	    return ret;
+	}
+    } else if (GMIME_IS_PART(me)) {
+	GMimePart* part = (GMimePart*)me;
+	GMimeDataWrapper* content = g_mime_part_get_content_object(part);
+	string type = g_mime_content_type_get_media_type(ct);
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	string charset;
+	const char* p = g_mime_object_get_content_type_parameter(me, "charset");
+	if (p) charset = g_mime_charset_canon_name(p);
+	if (type == "text") {
+	    string text;
+	    char buffer[SIZE];
+	    GMimeStream* sr = g_mime_data_wrapper_get_stream(content);
+	    auto encoding = g_mime_part_get_content_encoding(part);
+	    int len = 0;
+	    do {
+		len = g_mime_stream_read(sr, buffer, SIZE / 2);
+		if (len > 0) {
```

Comment:
I am not sure if I should use `SIZE/2` here.
If I am reading a base64 encode string and I full all the buffer (which seems unlikely, but not impossible), I will run out of memory when I try to decode that and store the result in the buffer.
I don't feel that using `SIZE/2` is a good solution, but I cannot come up with a better one.

Reply:
Why will you run out of memory?

The decoded version should be **smaller** - e.g.:

```
$ BASE64=c2ltcGxlIHRlc3QK
$ echo "$BASE64"|wc -c
17
$ echo c2ltcGxlIHRlc3QK|base64 --decode
simple test
$ echo c2ltcGxlIHRlc3QK|base64 --decode|wc -c
12
```

(The `wc -c` counts both include a linefeed character so are actually one more than the actual length.)

Reply:
Sorry, I think that the decoded version was bigger